### ISSUE TITLE
Change RWStructuredBuffer to StructuredBuffer where appropriate in Shaders

### ISSUE THREAD

**Issue by iamtrask** [OpenMined/OpenMined#244]
a few weeks ago we had a bug in our Compute Shaders where we got undefined behavior when we passed the same buffer into a kernel twice. Upon messaging Unity, today they got back to us with a very helpful response!!!

You're trying to use the same buffer as OUTPUT on 2 slots in the same compute shader. That is not a legal "move". For your case to work you should set them as read/input-only (use StructuredBuffer instead of RWStructuredBuffer in the shader for elems_a and elems_b) and it will work as expected.

Basically, we're using RWStructuredBuffer all over the place for variables that we aren't actually editing (RW == Read Write). We should swap out RWStructuredBuffer with StructuredBuffer for all of our buffers that we aren't editing :)
